### ISSUE TITLE
The column should be announced as “sortable” when screen reader is running, e.g., “Amount button – sortable”

### ISSUE THREAD

**Issue by supakitk** [gregnb/mui-datatables#875]
Please include suffix "sortable" for screen reader only user (don't need to be visible) when column is set to sortable.

**Comment by supakitk** [comment]
Thanks for the quick respond. Our accessibility team suggested an example at https://assets.cms.gov/resources/framework/2.0/Pages/datatables.html

When we have voice over on it will announce for example. “First name, sortable column. Activate to sort column ascending link”. 

The way they implement this is simply adding the alt text to the image. You can see from the code below. 

`<th scope="col" class="sorting_desc" role="columnheader" rowspan="1" colspan="1" abbr="First Name " aria-sort="descending" style="width: 150px;"><a class="sortingLink" href="#">First Name <img class="headerSortIcon" src="/resources/libs/datatables/1.9.1/images/sort_desc.png" alt="sortable column, sorted descending,  activate to sort column ascending"></a></th>`

Would this be something that possible to implement as well?

Thanks again.

**Comment by gabrielliwerant** [comment]
![R1fdEt3](https://user-images.githubusercontent.com/1779564/63980239-fa95e000-ca88-11e9-88fa-68f4da54b6ac.gif)

**Comment by gabrielliwerant** [comment]
Still trying to get a handle on the exact issue. Since we do have `aria-sort` once the sort for a header has been engaged, I'm guessing that the issue is when no sorting has yet been engaged, there is no signal that the header is sortable. In the example you show, there are always-present icons for sorting which allow you to use alt tags to present the information to a screen reader.

This table is following material ui, which doesn't have such a thing as always-present sortable icons at the moment. However, there is a tooltip with the `title="Sort"`. Is this not being read by your screen reader?

**Comment by supakitk** [comment]
Yes, that is not being read by the screen reader. You can try to use VoiceOver that comes with macOS (CMD+F5) and use keyboard to navigate to the column header. It is currently only announce the Header text.

**Comment by gabrielliwerant** [comment]
I use Chrome on a PC, and I just went looking for good accessibility tools, but it seems Chrome is not widely used by vision impaired folks.

In the mean time, do you think it would be sufficient to add `aria-sort="none"` on a sortable header prior to having a sort direction chosen? After choosing, it sets "descending" or "ascending" accordingly.

**Comment by patorjk** [comment]
@supakitk @gabrielliwerant I may be able to help as I use a Mac. I just tried VoiceOver (I've never used it before but it's pretty cool that it exists). In the current MUI setup, this is what VoiceOver reads to me when I tab over to the first column header in the "customize-search" example:

> Modified Title Label, button
> 
> Sort, you are currently on a button, inside of a cell. To click this button press Control+Option+Space.

The second part occurs because the tooltip pops up automatically because the field has focus. So for me, it's reading the tooltip. If I change the Tooltip's title from "Sort" to "Sort by Title", it reads me this:

> Sort by Title, you are currently on a button, inside of a cell. To click this button press Control+Option+Space.

I wonder if this is a browser issue. @supakitk what browser/version are you using?

Also, it looks like @Avd6977's PR https://github.com/gregnb/mui-datatables/pull/888 is related to this, but due to the # in the branch name it can't easily be checked with codesandbox. I tried it out on my Mac and it's setup to only give the "Sort by {label}" title if I change the textLabels.body.toolTip to be empty. I think the general idea is a pretty good, but it may make more sense to make textLabels.body.toolTip a template string with the `Sort by ${label}` value as a default.

**Comment by gabrielliwerant** [comment]
@patorjk The `title` is apparently non-standard for a lot of tags, so it's reasonable that certain browsers/screen readers would skip it.

The table is mostly material-ui underneath, so we handle sortable headers the way they do, and they do not have any sortable indication when there is no active sort. As far as I can tell, there is no accessibility requirement to announce that a table header is sortable or not. Seems reasonable to me to provide it, but I can't find any standard around it.

**Comment by Avd6977** [comment]
I'm not positive of the requirement, but it turns out it's partly linked to the #878 issue with using merge. For me, the titles were all blank because textLabels["body"]["toolTIp"] was undefined. However, just having it be sort does get flagged by our customers who run ADA checks because all of the titles are the same. Like I said, not sure of a requirement, but making them different/more clear I think would help.

**Comment by patorjk** [comment]
Doing a "grep -R toolTip src/*", it looks like that tooltip is only used in 1 place in the code (the column headers). One solution may be to create a textLabels.body.columnHeaderTooltip property which is a function. Ex:

```
columnHeaderTooltip: (label) => {
    return "Sort by " + label;
}
```

Then change:

```
<Tooltip
    title={options.textLabels.body.toolTip}
```

To:

```
<Tooltip
    title={options.textLabels.body.columnHeaderTooltip ?
        options.textLabels.body.columnHeaderTooltip(label) : options.textLabels.body.toolTip}
```

That would be backward compatible (though the label passing functionality from PR https://github.com/gregnb/mui-datatables/pull/888/ would also need to be brought in).

**Comment by supakitk** [comment]
@patorjk Sorry for the delay. I see the issue on both Safari version 13.0.1 and Chrome version 76 on macOS 10.14.6

**Comment by patorjk** [comment]
Does the problem still happen with version 2.10.1 of this library? I'm using macOS Mojave 10.14.6 and it reads the "Sort" tooltip text for me on Chrome 76 and Safari 12.1.2. 

Also, when I tab over to the header, as soon as the header has the focus, the tooltip pops up.

**Comment by Avd6977** [comment]
Sort is back after the 2.10.1 patch, but I would love to see it customizable like you'd listed above. This would now be a feature, though, not a bug.

**Comment by supakitk** [comment]
Hi @patorjk, Thanks! I can see it reads the "Sort" tooltip text after upgraded to version 2.10.1.

**Comment by gabrielliwerant** [comment]
Looks like the sort issue might have been related to the `textLabel`, though I'm not sure how! Glad it's fixed though.

@Avd6977 if you want to modify https://github.com/gregnb/mui-datatables/pull/888 to implement @patorjk's suggestion, I would accept it!

**Comment by Avd6977** [comment]
Updated it to reflect @patorjk's suggestion but I changed it to return column so someone could use name, id label, or whatever other property of column they want.

**Comment by gabrielliwerant** [comment]
Should be fixed as of `2.11.0`.
